### DISCUSSION TITLE
Time Attendance System

### DISCUSSION THREAD

**Discussion by dado-ca** [Budibase/budibase#13347] [Show and tell]
Hi there, we build an Employee Time Attendance app for our HR team, it uses PostgreSQL as backend database, and Jaspersoft Report Server Community Edition for printable Time Sheets. 

Features:
- Multiple shifts for different departments
- Employee belongs to one department only
- Automatically fill Time Attendance for all Employees belonging to specific department for selected time interval
- After Time Attendance has been added, HR manager adds different absences to Employees Time Sheet, like holiday, overtime, sick day and so on.
- Display Time Sheet for Employee, showing total time worked for selected period, and total time duration for other intervals (holiday, overtime, day off etc)
- Printable Time Sheets  created in Jaspersoft Report Server Community Edition

(Calendar would be nice, to show  time for users, tried full calender but id does not eork)
![ERV-Departments](https://github.com/Budibase/budibase/assets/44159556/a92ad396-7739-4c24-af0d-ac3a1bf068be)

Picture 1: Departments

![ERV-Epmloyees](https://github.com/Budibase/budibase/assets/44159556/3ae8dc60-a4f8-4e9f-b787-3f72b68316a9)

Picture 2: Employees

![ERV-Intervals](https://github.com/Budibase/budibase/assets/44159556/615ca1a5-a419-445b-adac-5f7f1e4445da)

Picture 3: Time intervals

![ERV-Shifts per Departments](https://github.com/Budibase/budibase/assets/44159556/6f8e4fd1-9dc2-4649-b604-c9a15ee09b98)

Picture 4: Time Shifts 

![ERV-AddTimeAttendanceByDepartment](https://github.com/Budibase/budibase/assets/44159556/96501d68-36b6-4497-884c-b6e38ab072a0)

Picture 5: Fill out  Time Attendance table for Employees beloging to specific department

![ERV-AddIntervalOneOrManyUSers](https://github.com/Budibase/budibase/assets/44159556/87041e50-7e63-4aa0-9273-ffbafc707a44)

Picture 6: Add new abscence, overtime etc, for one or many users

![ERV-EmployeeTimeSheet](https://github.com/Budibase/budibase/assets/44159556/8d4fa301-a3ec-45e4-8db8-0f5c96a45893)

Picture 7: Show Employees Time Sheet for selected time period

![ERV- TimeSheetJasperReports](https://github.com/Budibase/budibase/assets/44159556/4e5754ec-b3e4-4d68-bb10-acccd4b90325)

Picture 8: Printable Time Sheet (Jaspersoft)

![ERV-DeleteEmployeeTimeAttendance](https://github.com/Budibase/budibase/assets/44159556/14468868-5642-4169-912f-fa18f26ff677)

Picture 8: Delete Emploees Time Sheet

![ERV-SystemLog](https://github.com/Budibase/budibase/assets/44159556/89d4adfc-2aca-4f68-81d6-3703b294b916)

Picture 9: System logs

**Comment by turnerzzzz** [comment]
Wonderful work! We are trying to find alternatives for recording student attendance. Do you think this is applicable for more than 100 people? Is there any chance you could share the base template of this great work? I appreciate it!

**Reply by dado-ca** [reply]
Thanks, we are using it for 70 Employees, it runs on Postgres so it can handle a lot more. 

Here is the exported app and database schema:

[https://drive.google.com/drive/folders/1k-BEfADpNsuHJF-7-i3K4snjqpRP2Edh?usp=sharing](url)

Couple of things to consider:

In the Postgres table tHE_DayOfWeek you need to enter days of the week, starting from 0 for Sunday , 1 for Monday ... 6 for Saturday.

This is done, for the purpure of showing day of week short names in the timesheets on the frontend and on printed time sheet.

![Budibase TA day of week](https://github.com/user-attachments/assets/18ee6732-9279-4caa-b1a9-5f2dfef81e64)

Also, in the database schema, you will find procedures containing sections which pull days of week by numbers, and converting them to day of week short names. You need to change that into English.

On the frontend, on the Time Sheet screen (Šihterica), there are hardcoded calculations for every interval type you wish to view there. 

![TA Time sheet](https://github.com/user-attachments/assets/3853d635-c1c5-4eb9-bb67-f9e496d1aaca)

You define your own intervals on the Interval screen:

![TA Intervals](https://github.com/user-attachments/assets/c078d53f-4b6d-4e92-b8f5-b7bd2ee1733a)

Please notice,  I had make an mistake by updating budibase to the newest version. They change the way how Date/Time pickers handle dates and times. So it broke the whole logic of the app which take cares of work duration calculations! Very bad for a Time Attendance app. Those calculations are done via Postgres procedures. 

Maybe the app is not fully compatible with your use case, I hope you can pull some ideas from it.
